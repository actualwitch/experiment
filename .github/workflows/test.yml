name: Tests â–´ Unit

on:
  push:
    branches: ["main"]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup bun
        uses: oven-sh/setup-bun@v2
      - run: bun install
      - run: bun test src/feature
