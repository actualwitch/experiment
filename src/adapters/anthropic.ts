import type {
  MessageCreateParams,
  MessageCreateParamsNonStreaming,
  MessageParam,
  Tool,
} from "@anthropic-ai/sdk/resources/index.mjs";
import type { Message } from "../state/common";
import type { makeRequestTool } from "../state/inference";

export const experimentToAnthropic = (experiment: Message[]): MessageCreateParams | MessageCreateParamsNonStreaming => {
  let system = "";
  const messages: MessageParam[] = [];
  const tools: Tool[] = [];
  for (const { role, content } of experiment) {
    if (role === "system") {
      system += content + "\n";
    }
    if (role === "user") {
      if (typeof content === "string") {
        messages.push({ role: "user", content });
      }
      if (typeof content === "object") {
        messages.push({ role: "user", content: JSON.stringify(content) });
      }
    }
    if (role === "tool") {
      const thisTool = content;

      if (typeof thisTool === "object") {
        const tool = thisTool as typeof makeRequestTool;
        tools.push({
          name: tool?.function?.name ?? "Unnamed tool",
          description: tool?.function?.description ?? "No description",
          input_schema: tool?.function?.parameters,
        });
      }
    }
  }
  const tool_choice = tools.length === 1 ? { type: "tool" as const, name: tools[0].name } : undefined;
  return {
    system,
    messages,
    tools,
    tool_choice,
    model: "claude-3-5-haiku-20241022",
    temperature: 0.0,
    max_tokens: 2048,
    stream: true,
  };
};
