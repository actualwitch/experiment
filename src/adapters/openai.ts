import type {
  ChatCompletionCreateParams,
  ChatCompletionMessageParam,
  ChatCompletionTool,
} from "openai/resources/index.mjs";
import type { Message } from "../state/common";

export const experimentToOpenai = (experiment: Message[]): ChatCompletionCreateParams | null => {
  if (experiment.length === 0) {
    return null;
  }
  const messages: ChatCompletionMessageParam[] = [];
  const tools: ChatCompletionTool[] = [];
  for (const { role, content } of experiment) {
    if (role === "tool") {
      let thisTool = content;

      if (typeof thisTool === "object") {
        tools.push(thisTool as any);
      }
      continue;
    }
    if (typeof content === "string") {
      messages.push({ role, content });
    }
    if (typeof content === "object") {
      messages.push({ role: "user", content: JSON.stringify(content) });
    }
  }
  const result: ChatCompletionCreateParams = {
    messages,
    model: "gpt-4o",
    temperature: 0.0,
    max_tokens: 2048,
    stream: true,
  };
  if (tools.length) {
    result.tools = tools;
    result.tool_choice =
      tools.length === 1 ?
        {
          type: "function" as const,
          function: { name: tools[0].function.name },
        }
      : undefined;
  }
  return result;
};
