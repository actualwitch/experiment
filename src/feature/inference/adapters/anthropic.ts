import type {
  MessageCreateParams,
  MessageCreateParamsNonStreaming,
  MessageParam,
  Tool,
} from "@anthropic-ai/sdk/resources/index.mjs";
import type { Message } from "../../../types";
import { tokenLimit } from "../../../const";
import { createContextFromFiles, iterateDir } from "../../router/Explore";

export async function experimentToAnthropic(
  experiment: Message[],
  { max_tokens = tokenLimit }: { max_tokens?: number } = {},
): Promise<MessageCreateParams | MessageCreateParamsNonStreaming> {
  let system = "";
  const messages: MessageParam[] = [];
  const tools: Tool[] = [];
  for (const { role, content, fromServer } of experiment) {
    if (role === "system") {
      system += `${content}
`;
      continue;
    }
    if (role === "user" || role === "assistant") {
      if (typeof content === "object") {
        messages.push({ role, content: JSON.stringify(content) });
      }
      if (typeof content === "string") {
        messages.push({ role, content });
      }
      continue;
    }
    if (role === "context") {
      const directory = content.directory as string;
      const files = await iterateDir(directory);
      const context = await createContextFromFiles(files, directory);
      messages.push({ role: "user", content: context });
    }
    if (role === "tool" && !fromServer) {
      if (typeof content === "object") {
        const tool = content as any;
        tools.push({
          name: tool?.function?.name ?? "Unnamed tool",
          description: tool?.function?.description ?? "No description",
          input_schema: tool?.function?.parameters,
        });
      }
    }
  }
  const tool_choice = tools.length === 1 ? { type: "tool" as const, name: tools[0].name } : undefined;
  return {
    system,
    messages,
    tools,
    tool_choice,
    model: "claude-3-5-haiku-20241022",
    temperature: 0.0,
    max_tokens,
    stream: true,
  };
}
