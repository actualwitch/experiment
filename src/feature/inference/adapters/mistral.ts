import type { ChatCompletionStreamRequest } from "@mistralai/mistralai/models/components";
import { newLine, tokenLimit } from "../../../const";
import { type ExperimentFunction, type Message, ObjectOrStringType, StringType } from "../../../types";
import { createContextFromFiles, iterateDir } from "../../../utils/context";
import { experimentFunctionToTool, tryParseFunctionSchema } from "../function";
import type { InferenceConfig } from "../types";

export const experimentToMistral = async (
  experiment: Message[],
  { n_tokens, model, temperature }: InferenceConfig,
): Promise<ChatCompletionStreamRequest> => {
  const messages: ChatCompletionStreamRequest["messages"] = [];
  const tools: ChatCompletionStreamRequest["tools"] = [];
  for (const { role, content, fromServer, name, pronouns } of experiment) {
    if (role === "tool" && !fromServer) {
      if (typeof content === "object" && content !== null) {
        const tool = tryParseFunctionSchema(content as Record<string, unknown>).unwrapOr(content as ExperimentFunction);
        tools.push(experimentFunctionToTool(tool));
        continue;
      }
    }

    if (role === "context") {
      const directory = content.directory as string;
      const files = await iterateDir(directory);
      const context = await createContextFromFiles(files, directory);
      messages.push({ role: "user", content: context });
      continue;
    }
    if (typeof content === "string" && StringType.guard(role)) {
      const message = { role, content };
      if (role === "user") {
        const identity = pronouns ? `${name} (${pronouns})` : name;
        message.content = identity ? `${identity}:${newLine}${message.content}` : message.content;
      }
      messages.push(message);
      continue;
    }
    if (role !== "info" && ObjectOrStringType.guard(role)) {
      messages.push({ role, content: typeof content === "string" ? content : JSON.stringify(content) });
    }
  }
  const result: ChatCompletionStreamRequest = {
    messages,
    model,
    temperature,
    maxTokens: n_tokens,
    stream: true,
  };
  if (tools.length) {
    result.tools = tools;
    result.toolChoice =
      tools.length === 1
        ? {
            type: "function" as const,
            function: { name: tools[0].function.name },
          }
        : undefined;
  }
  return result;
};
