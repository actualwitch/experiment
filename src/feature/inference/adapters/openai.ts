import type {
  ChatCompletionCreateParams,
  ChatCompletionMessageParam,
  ChatCompletionTool,
} from "openai/resources/index.mjs";
import { type Message, ObjectOrStringType, StringType } from "../../../types";
import { createXMLContextFromFiles, iterateDir } from "../../../utils/context";

export const experimentToOpenai = async (experiment: Message[]): Promise<ChatCompletionCreateParams> => {
  const messages: ChatCompletionMessageParam[] = [];
  const tools: ChatCompletionTool[] = [];
  for (const { role, content, fromServer } of experiment) {
    if (role === "tool" && !fromServer) {
      const thisTool = content;

      if (typeof thisTool === "object") {
        tools.push(thisTool as any);
      }
      continue;
    }
    if (role === "context") {
      const directory = content.directory as string;
      const files = await iterateDir(directory);
      const context = await createXMLContextFromFiles(files, directory);
      messages.push({ role: "user", content: context });
      continue;
    }
    if (typeof content === "string" && StringType.guard(role)) {
      messages.push({ role, content });
      continue;
    }
    if (role !== "info" && ObjectOrStringType.guard(role)) {
      messages.push({ role, content: typeof content === "string" ? content : JSON.stringify(content) });
    }
  }
  const result: ChatCompletionCreateParams = {
    messages,
    model: "gpt-4o",
    temperature: 0.0,
    stream: true,
  };
  if (tools.length) {
    result.tools = tools;
    result.tool_choice = "auto";
  }
  return result;
};
