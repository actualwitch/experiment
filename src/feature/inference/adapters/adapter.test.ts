import { describe, expect, it } from "bun:test";
import { experimentToAnthropic } from "./anthropic";
import templates from "../../../../fixtures/templates.json";
import anthropicTool from "../../../../fixtures/tools/stockPrice.json";
import { experimentToOpenai } from "./openai";
import { experimentToMistral } from "./mistral";

for (const [provider, adapter] of [
  ["anthropic", experimentToAnthropic],
  ["mistral", experimentToMistral],
  ["openai", experimentToOpenai],
] as const) {
  describe(`inference/${provider}`, () => {
    it("tool use experiment", async () => {
      const result = await adapter(templates["Tool use"].messages);
      expect(result).toMatchSnapshot();
    });
    it("sample", async () => {
      const result = await adapter(templates["Sample chat"].messages);
      expect(result).toMatchSnapshot();
    });
    // it("anthropic tool format", async () => {
    //   const result = await adapter([{ role: "tool", content: anthropicTool }]);
    //   expect(result).toMatchSnapshot();
    // });
  });
}
