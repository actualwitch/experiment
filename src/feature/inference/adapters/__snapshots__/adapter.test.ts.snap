// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`inference/anthropic tool use experiment 1`] = `
{
  "max_tokens": 8192,
  "messages": [
    {
      "content": "hello friend! my name is ade, please remember that. then, search for the definition of singularity",
      "role": "user",
    },
  ],
  "model": "claude-3-5-haiku-20241022",
  "stream": true,
  "system": "",
  "temperature": 0,
  "tools": [
    {
      "description": "The \`bio\` tool allows you to persist information across conversations. The information will appear in the model set context below in future conversations.",
      "input_schema": {
        "properties": {
          "memory": {
            "description": "The information you want to persist across conversations.",
            "type": "string",
          },
        },
        "required": [
          "memory",
        ],
        "type": "object",
      },
      "name": "bio",
    },
    {
      "description": "The \`search\` tool allows you to search for information.",
      "input_schema": {
        "properties": {
          "query": {
            "description": "The information you want to search for.",
            "type": "string",
          },
        },
        "required": [
          "query",
        ],
        "type": "object",
      },
      "name": "search",
    },
  ],
}
`;

exports[`inference/anthropic sample 1`] = `
{
  "max_tokens": 8192,
  "messages": [
    {
      "content": "This is a user message.",
      "role": "user",
    },
    {
      "content": 
"
## Header

This is a small showcase of common elements in markdown for testing purposes. Here's a \`span\`, and *italics*  and **bold** text. Here's a [link](https://www.kaggle.com). Here's a list:

- item 1
- item 2
- item 3

Here's a numbered list:

1. item 1
2. item 2
3. item 3

Here's a table:

| Header 1 | Header 2 |
|----------|----------|
| cell 1   | cell 2   |
| cell 3   | cell 4   |

Here's a code block:

\`\`\`typescript
const interactive = css\`
  cursor: pointer;
  user-select: none;
  :hover {
    opacity: 1;
  }
\`;
const Emphasis = styled.em<{ isCollapsed?: boolean }>(
  css\`
    font-weight: italic;
    opacity: 0.7;
  \`,
  ({ isCollapsed }) => {
    if (!isCollapsed) return;
    return css\`
      :before {
        content: "( ";
        opacity: 0.3;
      }
      :after {
        content: " )";
        opacity: 0.3;
      }
    \`;
  },
  interactive,
);
\`\`\`

Here's an image:

![image](https://www.kaggle.com/static/images/site-logo.png)

Here's a blockquote:

> This is a blockquote.

Here's a horizontal rule:

---
"
,
      "role": "assistant",
    },
  ],
  "model": "claude-3-5-haiku-20241022",
  "stream": true,
  "system": 
"This is a system message.
"
,
  "temperature": 0,
  "tools": [],
}
`;

exports[`inference/mistral tool use experiment 1`] = `
{
  "maxTokens": 8192,
  "messages": [
    {
      "content": "hello friend! my name is ade, please remember that. then, search for the definition of singularity",
      "role": "user",
    },
    {
      "content": "{"function":"bio","arguments":{"memory":"User's name is Ade."}}",
      "role": "tool",
    },
    {
      "content": "{"function":"search","arguments":{"query":"definition of singularity"}}",
      "role": "tool",
    },
  ],
  "model": "mistral-small-latest",
  "stream": true,
  "temperature": 0,
  "toolChoice": undefined,
  "tools": [
    {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "function": {
        "description": "The \`bio\` tool allows you to persist information across conversations. The information will appear in the model set context below in future conversations.",
        "name": "bio",
        "parameters": {
          "properties": {
            "memory": {
              "description": "The information you want to persist across conversations.",
              "type": "string",
            },
          },
          "required": [
            "memory",
          ],
          "type": "object",
        },
      },
      "type": "function",
    },
    {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "function": {
        "description": "The \`search\` tool allows you to search for information.",
        "name": "search",
        "parameters": {
          "properties": {
            "query": {
              "description": "The information you want to search for.",
              "type": "string",
            },
          },
          "required": [
            "query",
          ],
          "type": "object",
        },
      },
      "type": "function",
    },
  ],
}
`;

exports[`inference/mistral sample 1`] = `
{
  "maxTokens": 8192,
  "messages": [
    {
      "content": "This is a system message.",
      "role": "system",
    },
    {
      "content": "This is a developer message.",
      "role": "developer",
    },
    {
      "content": "This is a user message.",
      "role": "user",
    },
    {
      "content": 
"
## Header

This is a small showcase of common elements in markdown for testing purposes. Here's a \`span\`, and *italics*  and **bold** text. Here's a [link](https://www.kaggle.com). Here's a list:

- item 1
- item 2
- item 3

Here's a numbered list:

1. item 1
2. item 2
3. item 3

Here's a table:

| Header 1 | Header 2 |
|----------|----------|
| cell 1   | cell 2   |
| cell 3   | cell 4   |

Here's a code block:

\`\`\`typescript
const interactive = css\`
  cursor: pointer;
  user-select: none;
  :hover {
    opacity: 1;
  }
\`;
const Emphasis = styled.em<{ isCollapsed?: boolean }>(
  css\`
    font-weight: italic;
    opacity: 0.7;
  \`,
  ({ isCollapsed }) => {
    if (!isCollapsed) return;
    return css\`
      :before {
        content: "( ";
        opacity: 0.3;
      }
      :after {
        content: " )";
        opacity: 0.3;
      }
    \`;
  },
  interactive,
);
\`\`\`

Here's an image:

![image](https://www.kaggle.com/static/images/site-logo.png)

Here's a blockquote:

> This is a blockquote.

Here's a horizontal rule:

---
"
,
      "role": "assistant",
    },
  ],
  "model": "mistral-small-latest",
  "stream": true,
  "temperature": 0,
}
`;

exports[`inference/openai tool use experiment 1`] = `
{
  "max_tokens": 8192,
  "messages": [
    {
      "content": "hello friend! my name is ade, please remember that. then, search for the definition of singularity",
      "role": "user",
    },
    {
      "content": "{"function":"bio","arguments":{"memory":"User's name is Ade."}}",
      "role": "tool",
    },
    {
      "content": "{"function":"search","arguments":{"query":"definition of singularity"}}",
      "role": "tool",
    },
  ],
  "model": "gpt-4o",
  "stream": true,
  "stream_options": {
    "include_usage": true,
  },
  "temperature": 0,
  "tool_choice": "auto",
  "tools": [
    {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "function": {
        "description": "The \`bio\` tool allows you to persist information across conversations. The information will appear in the model set context below in future conversations.",
        "name": "bio",
        "parameters": {
          "properties": {
            "memory": {
              "description": "The information you want to persist across conversations.",
              "type": "string",
            },
          },
          "required": [
            "memory",
          ],
          "type": "object",
        },
      },
      "type": "function",
    },
    {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "function": {
        "description": "The \`search\` tool allows you to search for information.",
        "name": "search",
        "parameters": {
          "properties": {
            "query": {
              "description": "The information you want to search for.",
              "type": "string",
            },
          },
          "required": [
            "query",
          ],
          "type": "object",
        },
      },
      "type": "function",
    },
  ],
}
`;

exports[`inference/openai sample 1`] = `
{
  "max_tokens": 8192,
  "messages": [
    {
      "content": "This is a system message.",
      "role": "system",
    },
    {
      "content": "This is a developer message.",
      "role": "developer",
    },
    {
      "content": "This is a user message.",
      "role": "user",
    },
    {
      "content": 
"
## Header

This is a small showcase of common elements in markdown for testing purposes. Here's a \`span\`, and *italics*  and **bold** text. Here's a [link](https://www.kaggle.com). Here's a list:

- item 1
- item 2
- item 3

Here's a numbered list:

1. item 1
2. item 2
3. item 3

Here's a table:

| Header 1 | Header 2 |
|----------|----------|
| cell 1   | cell 2   |
| cell 3   | cell 4   |

Here's a code block:

\`\`\`typescript
const interactive = css\`
  cursor: pointer;
  user-select: none;
  :hover {
    opacity: 1;
  }
\`;
const Emphasis = styled.em<{ isCollapsed?: boolean }>(
  css\`
    font-weight: italic;
    opacity: 0.7;
  \`,
  ({ isCollapsed }) => {
    if (!isCollapsed) return;
    return css\`
      :before {
        content: "( ";
        opacity: 0.3;
      }
      :after {
        content: " )";
        opacity: 0.3;
      }
    \`;
  },
  interactive,
);
\`\`\`

Here's an image:

![image](https://www.kaggle.com/static/images/site-logo.png)

Here's a blockquote:

> This is a blockquote.

Here's a horizontal rule:

---
"
,
      "role": "assistant",
    },
  ],
  "model": "gpt-4o",
  "stream": true,
  "stream_options": {
    "include_usage": true,
  },
  "temperature": 0,
}
`;
